sqld/src/database/write_proxy.rs

```diff
@@ -115,50 +115,67 @@ impl WriteProxyDatabase {
+                            RowResult::Error(e) => Some(Err(Error::RpcQueryError(e))),
+                        }
+                    })
+                    .collect();
+
+                Ok((results, *state))
+            }
+            Err(e) => {
+                // Set state to invalid, so next call is sent to remote, and we have a chance
+                // to recover state.
+                *state = State::Invalid;
+                Err(Error::RpcQueryExecutionError(e))
+            }
+        }
+    }
 }
 
 #[async_trait::async_trait]
 impl Database for WriteProxyDatabase {
     async fn execute_program(&self, pgm: Program) -> Result<(Vec<Option<QueryResult>>, State)> {
         let mut state = self.state.lock().await;
-        if *state == State::Init
-            && pgm.is_read_only()
-            && final_state(*state, pgm.steps.iter().map(|s| &s.query.stmt)) == State::Init
-        {
-            self.read_db.execute_program(pgm).await
-        } else {
-            let mut client = self.write_proxy.clone();
-            let req = crate::rpc::proxy::rpc::ProgramReq {
-                client_id: self.client_id.to_string(),
-                pgm: Some(pgm.into()),
-            };
-            match client.execute(req).await {
-                Ok(r) => {
-                    let execute_result = r.into_inner();
-                    *state = execute_result.state().into();
-                    let results = execute_result
-                        .results
-                        .into_iter()
-                        .map(|r| -> Option<QueryResult> {
-                            let result = r.row_result?;
-                            match result {
-                                RowResult::Row(res) => {
-                                    Some(Ok(QueryResponse::ResultSet(res.into())))
-                                }
-                                RowResult::Error(e) => Some(Err(Error::RpcQueryError(e))),
-                            }
-                        })
-                        .collect();
-
-                    Ok((results, *state))
-                }
-                Err(e) => {
-                    // Set state to invalid, so next call is sent to remote, and we have a chance
-                    // to recover state.
-                    *state = State::Invalid;
-                    Err(Error::RpcQueryExecutionError(e))
+        if *state == State::Init && pgm.is_read_only() {
+            // This program won't perform any write, we try to predict it's final state, i.e, if
+            // it's an interactive transaction or not, and then attempt to execute it. If it turns
+            // out that it was indeed an interactive transaction, we rollback the current
+            // transaction, and attempt to execute again on the remote
+            if predict_final_state(*state, pgm.steps.iter().map(|s| &s.query.stmt)) == State::Init {
```

Comment:
I think that this check will always fail for batches that are produced by libsql client. The batch looks like `BEGIN`, ..., `COMMIT`, `ROLLBACK`, so the `predict_final_state()` function will always return `State::Invalid`, right?

Reply:
If we verify that all statements are read-only, then perhaps we can always just execute the batch locally, and if this leaves a transaction open, then rollback it and retry on master? In effect, do the same as now, without checking for the predicted final state.